libs/domain/checkout/src/services/default-checkout.service.ts

```diff
@@ -1,140 +1,172 @@
+          this.state.pipe(
+            map((state) =>
+              isEmpty ? CheckoutProcessState.NotAvailable : state
+            )
+          )
+        )
+      );
   }
 
   placeOrder(): Observable<CheckoutResponse> {
+    this.state.next(CheckoutProcessState.Busy);
     return subscribeReplay(
-      this.orchestrationService.submit().pipe(
-        map((validity) =>
-          validity.every(({ validity }) => validity === Validity.Valid)
-        ),
-        switchMap((canCheckout) =>
-          canCheckout
-            ? this.preparePayload()
-            : throwError(
-                () =>
-                  new Error(
-                    'Cannot checkout, check your cart value limits and validations'
-                  )
-              )
-        ),
-        switchMap((payload) =>
-          this.adapter.placeOrder({ attributes: payload })
-        ),
+      this.collect().pipe(
+        catchError((error) => {
+          this.state.next(CheckoutProcessState.Ready);
+          return EMPTY;
+        }),
+        switchMap((data: Checkout) => {
+          const attributes = this.amendCheckoutData(data);
+          return this.adapter.placeOrder({ attributes }).pipe(
+            finalize(() => {
+              this.state.next(CheckoutProcessState.Ready);
+              return EMPTY;
+            })
+          );
+        }),
+        switchMap(this.resolveRedirect),
         tap((response) => this.postCheckout(response))
       )
     );
   }
 
-  protected getCustomer(): Observable<ContactDetails | null> {
-    return concat(
-      this.dataService.getCustomer().pipe(take(1), filter(Boolean)),
-      // TODO: Workaround for the case when customer is not set in the data service
-      // Proper implementation should get customer data from the UserService.
-      of({
-        email: 'temporary-email@temporary-workaround.com',
-      })
-    ).pipe(take(1));
-  }
-
-  protected preparePayload(): Observable<Checkout> {
-    return combineLatest([
-      this.cartService.getCart().pipe(map((cart) => cart?.id)),
-      this.getCustomer(),
-      this.dataService.getAddress(),
-      this.dataService.getShipment(),
-      this.dataService.getPayment(),
-    ]).pipe(
-      take(1),
-      map(([cartId, customer, billingAddress, shipment, payment]) => {
-        const payload: Checkout = {
-          cartId: cartId!,
-          customer: {
-            ...customer!,
-            salutation:
-              customer?.salutation ?? billingAddress!.salutation ?? '',
-          },
-          billingAddress: {
-            ...billingAddress!,
-          },
-        };
+  protected collect(): Observable<T> {
+    if (this.steps.length === 0) {
+      throw new Error('No steps registered');
+    }
 
-        if (shipment?.idShipmentMethod) {
-          payload.shipment = { idShipmentMethod: shipment.idShipmentMethod };
-          payload.shippingAddress = { ...billingAddress! };
-        }
+    const collectStep = (
```

Comment:
This recursion looks weird

Reply:
Have left a comment in the code to clarify.

Reply:
I agree it's complex, I'd try to replace it with either for loop or `concat` + `filter`.

On the other hand, we will still need to observe all steps, as they may become invalid at unpredictable time.

Reply:
We also want to trigger validity checks for all components on single page checkout (to mark all invalid fields in all steps, not only half of it)

Reply:
No we do not, we like to only mark invalid fields step by step on single page as well, as otherwise the errors look to overwhelming. It might be opinionated and we could add an option to validate it all.